### ISSUE TITLE
[JAVA] XPath rules don't catch error cases found in Designer.

### ISSUE THREAD

**Issue by tjoneslo** [pmd/pmd#260]
I'm trying to build a custom rule to detect if specific variables declared as `PropertyKey` are always declared as `public static final`. I am using version 5.5.2, following the tutorial, used the designer to create the XPath rules, and verified they are correct on real code. But when running through Gradle on the whole project, PMD doesn't flag the errors.  

The rule as written: 
```
    <rule name="PropertyKeys must be public static final"
          message="ProperyKey definition must be public static final"
          class="net.sourceforge.pmd.lang.rule.XPathRule"
          language="java">
        <description>
            PropertyKey definition must be public static final
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
        //FieldDeclaration[./Type/ReferenceType/ClassOrInterfaceType[typeof(@Image, 'mme.foundation.properties', 'PropertyKey')] and
        (@Static='false' or @Final='false' or @Public='false') ]
                </value>
            </property>
        </properties>
    </rule>
```

And an example failure case not found:

```
package foo;

import mme.foundation.PropertyDefs;
import mme.foundation.properties.PropertyKey;

    @ProviderFor(PropertyDefs.class)
    public static class EndpointProps extends PropertyDefs {
        private static final PropertyKey<Map<String, String>> ServiceKey =
                PropertyKey.Map(
                        "KeyName",
                        ImmutableMap.of(
                                "endpoint/key", "endpoint/value"));
    }
```
**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Gradle

**Comment by jsotuyod** [comment]
The rule seems to be correct. Gradle populates the auxclasspath properly (provided you are using a somewhat current version).

Which Gradle version are you using?
Can you submit the configuration for PMD in your gradle build?

**Comment by tjoneslo** [comment]
```
$ ./gradlew -v

------------------------------------------------------------
Gradle 3.2.1
------------------------------------------------------------

Build time:   2016-11-22 15:19:54 UTC
Revision:     83b485b914fd4f335ad0e66af9d14aad458d2cc5

Groovy:       2.4.7
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_111 (Oracle Corporation 25.111-b14)
OS:           Mac OS X 10.12.4 x86_64
```

Configuration for PMD

```
apply plugin: 'pmd'
pmd {
    ignoreFailures = true
    toolVersion = '5.5.2'
    ruleSetFiles = files(rootProject.file('gradle/codequality/pmd-rules.xml'))
    ruleSets = []
    consoleOutput = true
}

tasks.withType(Pmd) { t ->
    t.exclude '**/generated/**'
    reports.xml.enabled = true
}
```

**Comment by jsotuyod** [comment]
It's working on my sample project created by just copy / pasting the info you provided.... Maybe something wrong in your `pmd-rules.xml` file?

For your reference, I here provide my working sample bassed on your info: [sample.zip](https://github.com/pmd/pmd/files/775352/sample.zip)

```
$ gradle pmdMain
:pmdMain
* file: ./src/main/java/Test.java
    src:  Test.java:4:4n
    rule: PropertyKeys must be public static final
    msg:  ProperyKey definition must be public static final
    code: public PropertyKey<Map<String, String>> OtherKey = null;

    src:  Test.java:7:7
    rule: PropertyKeys must be public static final
    msg:  ProperyKey definition must be public static final
    code: private static final PropertyKey<Map<String, String>> ServiceKey = null;



Summary:

 : 2
* warnings: 2
2 PMD rule violations were found. See the report at: file:///home/jmsotuyo/monits/test/build/reports/pmd/main.html

BUILD SUCCESSFUL

Total time: 13.435 secs
```

**Comment by tjoneslo** [comment]
Here's my complete `pmd-rules.xml` file as used by the project. I can't see anything thing that might cause an issue. 

```
<?xml version="1.0"?>
<ruleset name="ruleset"
         xmlns="http://pmd.sf.net/ruleset/1.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd"
         xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd">

    <description>
       PMD ruleset
    </description>

    <rule ref="rulesets/java/basic.xml"/>

    <rule ref="rulesets/java/logging-java.xml">
        <exclude name="GuardLogStatementJavaUtil" />
        <exclude name="LoggerIsNotStaticFinal" />
    </rule>

    <rule ref="rulesets/java/braces.xml/WhileLoopsMustUseBraces"/>
    <rule ref="rulesets/java/braces.xml/IfElseStmtsMustUseBraces"/>
    <rule ref="rulesets/java/braces.xml/ForLoopsMustUseBraces"/>

    <rule ref="rulesets/java/imports.xml/DuplicateImports"/>
    <rule ref="rulesets/java/imports.xml/DontImportJavaLang"/>
    <rule ref="rulesets/java/imports.xml/ImportFromSamePackage"/>
    <rule ref="rulesets/java/imports.xml/UnnecessaryFullyQualifiedName"/>

    <rule ref="rulesets/java/unnecessary.xml">
        <exclude name="UselessParentheses"/>
    </rule>
    <rule ref="rulesets/java/empty.xml">
        <!-- Customized below. -->
        <exclude name="EmptyCatchBlock"/>
    </rule>
    <rule ref="rulesets/java/empty.xml/EmptyCatchBlock"
            message="Avoid uncommented empty catch blocks.">

        <properties>
            <property name="allowCommentedBlocks" value="true"/>
        </properties>
    </rule>
    <!--<rule ref="rulesets/java/comments.xml"/>-->

    <rule name="PropertyKeys must be public static final"
          message="ProperyKey definition must be public static final"
          class="net.sourceforge.pmd.lang.rule.XPathRule"
          language="java">
        <description>
            PropertyKey definition must be public static final
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
        //FieldDeclaration[./Type/ReferenceType/ClassOrInterfaceType[typeof(@Image, 'mme.foundation.properties', 'PropertyKey')] and
        (@Static='false' or @Final='false' or @Public='false') ]
                </value>
            </property>
        </properties>
    </rule>
</ruleset>
```

**Comment by jsotuyod** [comment]
@tjoneslo maybe runnning `./gradlew pmdMain --info` will give us some insight as to what my be going on?

**Comment by tjoneslo** [comment]
If I pull all of the rules other than the specific one in question out of the `pmd-rules.xml` file, it now works, as in it fails with the PMD error. Let me work on being selective about which rules I do have are interfering with the xpath one.

**Comment by jsotuyod** [comment]
@tjoneslo please, let me know of your findings. I'm most interested on knowing what's going on here.

**Comment by tjoneslo** [comment]
I've found that adding any or all of the following rules refs to the `pmd-rules.xml` causes it to stop processing the xroute rules.  

```
    <rule ref="rulesets/java/basic.xml"/>
    <rule ref="rulesets/java/logging-java.xml"/>
   <rule ref="rulesets/java/imports.xml/DuplicateImports"/>
   <rule ref="rulesets/java/unnecessary.xml"/>
```

but  adding either of these still works correctly. 

```
<rule ref="rulesets/java/braces.xml/>
<rule ref="rulesets/java/empty.xml"/>
```

I'll need to dig further into these to find why

**Comment by adangel** [comment]
Your "PropertyKeys" rule uses the `typeof` xpath function. Reading the code, it takes a "fullTypeName" and optional a "shortTypeName". You are specifying just the package name as the fullTypeName. Please try using the fully qualified class name, like this:

    typeof(@Image, 'mme.foundation.properties.PropertyKey', 'PropertyKey')

Depending on whether the auxclasspath option is used or not, this might make a difference. I think, executing the rule with the Designer tool, doesn't make use of the auxclasspath and it simply matched the "shortTypeName" via a string comparison.

Why it works differently depending on different rules activated, I don't know...

For reference, here's the implementation: [TypeOfFunction.java](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/xpath/TypeOfFunction.java)

**Comment by jsotuyod** [comment]
@tjoneslo @adangel is right, I hadn't noticed before.

Changing the rule a explained should get it working regardless of other rules. The problem is triggered by those rulesets is that they ask PMD to perform type resolution, which your rule uses, but doesn't require itself.

To be absolutely correct, your rule should be stated as:

```xml
    <rule name="PropertyKeys must be public static final"
          message="ProperyKey definition must be public static final"
          class="net.sourceforge.pmd.lang.rule.XPathRule"
          language="java"
          typeResolution="true">
        <description>
            PropertyKey definition must be public static final
        </description>
        <priority>3</priority>
        <properties>
            <property name="xpath">
                <value>
        //FieldDeclaration[./Type/ReferenceType/ClassOrInterfaceType[typeof(@Image, 'mme.foundation.properties.PropertyKey', 'PropertyKey')] and
        (@Static='false' or @Final='false' or @Public='false') ]
                </value>
            </property>
        </properties>
    </rule>
```

Notice both, the `typeResolution="true"` attribute, and the changes to `typeof()` explained by @adangel

**Comment by tjoneslo** [comment]
Yes, that works much better, thank you. I've lost the references to where I saw the `typeof()` documentation.

**Comment by jsotuyod** [comment]
@tjoneslo glad to hear. I'm closing this issue, I'll add a task to my backlog to review existing docs on the web for building xpath rules with type resolution.

**Comment by adangel** [comment]
Great! I didn't remember the `typeResolution="true"`, but now it makes perfect sense.
About the documentation: It's probably missing, or at least incomplete... https://pmd.github.io/pmd-5.5.3/customizing/howtowritearule.html#Inside_an_XPath_query 
it would benefit from being a bit more accurate :-)
